src/ucx_plugin.c

```diff
@@ -276,114 +278,123 @@ static ncclResult_t ucx_worker_get_netaddress(ucp_worker_h worker,
 }
 
 static ncclResult_t ucx_get_ctx_and_worker(int dev, ucp_context_h *ctx,
-                                           ucp_worker_h *worker,
+                                           nccl_ucx_worker_t **ucx_worker,
                                            ucp_tag_t *newtag) {
   pthread_mutex_lock(&nccl_ucx_lock);
-  if (ncclNIbDevs < dev) {
+  if (ncclNIbDevs <= dev) {
     WARN("Device index is too large");
-    return ncclSystemError;
+    goto err;
   }
 
-  if (workers[dev].count == 0) {
-    ucx_init_context(&workers[dev].ctx, dev);
-    ucx_init_worker(workers[dev].ctx, &workers[dev].worker);
-    workers[dev].last_tag = tag;
-    worker_count++;
+  nccl_ucx_worker_t *w;
+  for (w = workers[dev]; w != NULL; w = w->next) {
+    if (w->dev == dev && w->thread == pthread_self()) {
+      break;
+    }
   }
 
-  *ctx    = workers[dev].ctx;
-  *worker = workers[dev].worker;
+  if (w == NULL) {
+    w = calloc(1, sizeof(*w));
+    if (w == NULL) {
+      WARN("Worker allocation failure");
+      goto err;
+    }
+
+    w->dev    = dev;
+    w->thread = pthread_self();
+    w->count  = 0;
 
+    ucx_init_context(&w->ctx, dev);
```

Comment:
i think we do not need a separate context per worker on the same device

Reply:
agree, since that is not introduced by this PR, I will address it in the next PR (and will sanity check perf), if ok for you

Reply:
next pr is fine by me, but pls note that previous design was to have a single context+worker per device.And that worker could be shared by multiple threads.
Now we may have several workers for the same device and they could have the same contetx

Reply:
now i see your point, it's actually introduced by this PR. i'll prepare another PR for that.